sourcegraph/sourcegraph
# web: upgrade Storybook

## Context

Upgrading Storybook to the latest version to benefit from [goodies released](https://github.com/storybookjs/storybook/blob/next/CHANGELOG.md) since [Oct 14](https://www.npmjs.com/package/@storybook/core/v/6.3.12) and eventually [enable lazy compilation in development mode](https://github.com/sourcegraph/sourcegraph/issues/36438). Read more about Storybook Lazy Compilation for Webpack [here](https://storybook.js.org/blog/storybook-lazy-compilation-for-webpack/).

[Slack thread](https://sourcegraph.slack.com/archives/C01LTKUHRL3/p1654084573922929) for context.

## Changes

1. Replaced [deprecated static directories CLI flag](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated---static-dir-cli-flag) with the configuration field.
2. Replaced [ChromaticStory](https://github.com/sourcegraph/sourcegraph/pull/36437/files#diff-73da9bb4eedaef0ca9730e2d30cca826df1c2191f7fad8af37140d79c6de8298) with [withChromaticThemes](https://github.com/sourcegraph/sourcegraph/pull/36437/files#diff-b142a47a403a3678de9e3cab3fa099c598bd0f4471682041b97d17fb67217709) global decorator to snapshot stories in the dark mode in Chromatic automatically. This is necessary because of [the story-store API change](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#story-store-v7) that we relied on.
3. Introduced our wrapper around the `isChromatic` helper to locally test the change described above. Run `CHROMATIC=true yarn storybook` to test it.
4. Pinned `prettier` version to avoid huge diff in this PR. Let's unpin it in a follow-up PR.  
5. Added React context to the `Popover` component to be able to change the HTML element where it's rendered. Thanks, @vovakulikov, for implementing that! 🙌

### New global decorator

The [withChromaticThemes](https://github.com/sourcegraph/sourcegraph/pull/36437/files#diff-b142a47a403a3678de9e3cab3fa099c598bd0f4471682041b97d17fb67217709) decorator renders stories with `parameters.chromatic.enableDarkMode === true` twice — in light and dark mode. The decorator is enabled only if `isChromatic() === true`. It's a [suggested](https://www.chromatic.com/docs/faq#do-you-support-taking-snapshots-of-a-component-with-multiple-the) way of snapshotting the same story with different themes. Trade-offs:

#### Pros

1. Simpler code — decorator React component. No need to touch the story store API used previously.
2. Compatible with the new story store API. We won't need to migrate anything when the Storybook v7 is released.
3. Two themes on one snapshot —> less number of snapshots -> less spending.

#### Cons

1. Dark mode and light mode stories are not isolated — they are rendered in the same React tree. (e.g., that's why we need a way to control the `Popover` root element).

## Test plan

1. Ensure that Storybook works locally without changes.
2. Ensure that Chromatic snapshots are as expected.

## App preview:

- [Web](https://sg-web-vb-storybook-upgrade.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-uabvmwcqlt.chromatic.com)

To learn more, check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews).
